components/cdc/src/endpoint.rs

```diff
@@ -1044,58 +1015,86 @@ struct Initializer {
+        let sched = self.sched.clone();
+        let region_epoch = self.region_epoch.clone();
+        let downstream_id = self.downstream_id;
+        let downstream_state = self.downstream_state.clone();
+        let (cb, fut) = tikv_util::future::paired_future_callback();
+        if let Err(e) = raft_router.significant_send(
+            self.region_id,
+            SignificantMsg::CaptureChange {
+                cmd: change_cmd,
+                region_epoch,
+                callback: Callback::Read(Box::new(move |resp| {
+                    if let Err(e) = sched.schedule(Task::InitDownstream {
+                        downstream_id,
+                        downstream_state,
+                        cb: Box::new(move || {
+                            cb(resp);
+                        }),
+                    }) {
+                        error!("cdc schedule cdc task failed"; "error" => ?e);
+                    }
+                })),
+            },
+        ) {
+            warn!("cdc send capture change cmd failed";
+            "region_id" => self.region_id, "error" => ?e);
+            return Err(Error::request(e.into()));
+        }
+
+        match fut.await {
+            Ok(resp) => self.on_change_cmd_response(resp).await,
+            Err(e) => Err(Error::Other(box_err!(e))),
+        }
+    }
+
+    async fn on_change_cmd_response(
+        &mut self,
+        mut resp: ReadResponse<RocksSnapshot>,
+    ) -> Result<()> {
+        if let Some(region_snapshot) = resp.snapshot {
             assert_eq!(self.region_id, region_snapshot.get_region().get_id());
             let region = region_snapshot.get_region().clone();
             // Require barrier before finishing incremental scan, because
             // CDC needs to make sure resovled ts events can only be sent after
             // incremental scan is finished.
             let require_barrier = true;
-
-            let res = self
-                .async_incremental_scan(region_snapshot, region, require_barrier)
-                .await;
-
-            if res.is_ok() {
-                CDC_SCAN_TASKS.with_label_values(&["finish"]).inc();
-                return;
-            }
-
-            CDC_SCAN_TASKS.with_label_values(&["abort"]).inc();
-            let err = res.unwrap_err();
-            // Deregister downstream if incremental scan fails.
-            if self.build_resolver {
-                Deregister::Region {
-                    region_id: self.region_id,
-                    observe_id: self.observe_id,
-                    err,
-                }
-            } else {
-                Deregister::Downstream {
-                    region_id: self.region_id,
-                    downstream_id: self.downstream_id,
-                    conn_id: self.conn_id,
-                    err: Some(err),
-                }
-            }
+            self.async_incremental_scan(region_snapshot, region, require_barrier)
```

Comment:
Does `require_barrier` have to be true here?

Reply:
Yes, in the current implementation it has be to true. It's false only in tests that do not care about the barrier. This parameter can be removed after the segmentation scan is supported.

Reply:
> Yes, in the current implementation it has be to true. It's false only in tests that do not care about the barrier. This parameter can be removed after the segmentation scan is supported.

Ok. Can you add a TODO?